[FIX] Pre Calculates the crate animation math to fix the perfomance issues breaking crate animation

MNarath1 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

The animation math for crate was a bit to extensive for byond inside a for loop so i moved it out into an extra proc.
Said proc creates a assoc list for all the needed animation values with a unique value as the index means you can have crates with the same animation parameter share a list while crates with different parameter share another list and so on.
This allows for using pre calculated lists but holds the amount of lists to a minimum while still allowing for a great flexibility and easy crate implementation with new parameter.
Also thanks to park66665 for helping me out with the assoc list.
## Why It's Good For The Game

Fixes currently broken looking crate animation

## Changelog
:cl:

fix: crate animation looking broken
code: Move animation math into a extra proc to pre calculate it
refactor: Adding a association list to allow easy adding of additional crates with unique values while still keeping the amount of actual lists to a minimum.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

MNarath1 committed:
Pre calculates crate animation math

MNarath1 committed:
oops

park66665 commented on code/game/objects/structures/crates_lockers/crates.dm:
```diff
@@ -23,9 +23,11 @@
 	drag_slowdown = 0
 	var/obj/item/paper/fluff/jobs/cargo/manifest/manifest
 	var/radius_2 = 1.35
+	var/animation_math
```
```suggestion
	var/static/list/animation_math
```

MNarath1 replied:
I understand what you mean only problem is this would bring some problems for any crate that you want to have other values for things like the animation angle or azimuth angle or another animation time as those would share the list for items with these specific values now this is no problem at the moment but could introduce some headache later honestly i do not like the use of lists inside lists anyways i would more like to use a matrix instead of a list

MNarath1 replied:
but i think i can reduced the total ammount of lists generated to a single one for every obj without reducing the ammount of actual values so at least no lists inside lists

MNarath1 replied:
Wait i have an idea that might reduce the amount of lists more if i can create a static list for every child obj of crates so that all child objects of the same type share that list but i am not sure how to best go about that

MNarath1 replied:
i think i can pull this off i will create a static list and for every child object a list inside with the first index beeing used to identify its obj path thanks for that idea its still a few more lists but it will greatly reduce the total amount of actual lists

park66665 replied:
If you feel *really* brave, make it a static assoc list of lists with animation-determining parameters as a key so every crate with the same animation parameters can share a single pre-calculated animation list, e.g.:
`animation_list["[door_anim_time]-[door_anim_angle]-[door_hinge]-[radius_2]-[open_angle]"][idx]`

MNarath1 replied:
Yes did it thanks for that idea again got a assoziative list that gives back the needed index for the animation_math list and thus only creates a new list if there is actually a crate that gets initialized with a new set of values(or if an admin vv's a crate and calls the proc)

park66665 commented on code/game/objects/structures/crates_lockers/crates.dm:
```diff
@@ -23,9 +23,11 @@
 
 /obj/structure/closet/crate/Initialize()
 	. = ..()
+	animation_list()
```
```suggestion
	if(!animation_math)
		animation_list()
```

park66665 reviewed (changes requested):
Creating a list of lists for each object is not a good idea. You can use static lists to reduce the number of lists of lists.

MNarath1 committed:
reducing lists to one extra list per obj

While a static list would allow this to become one global list for every crate it would make it a bit harder to make any exceptions for crates that need other values but reducing the ammounts of lists is still a good idea so i am reducing it  to a single list per object still wish i could just properly define an actual array for thing like this.

MNarath1 committed:
no reason to make a list if it has no animation

MNarath1 committed:
just small change to make code more flexible

MNarath1 committed:
oops missspelled comment

MNarath1 committed:
Adds in a associative list to get list index and reduce the lists needed

Now this adds an associative static list to get the list index for the main list also checks if there is already a list for the specific animation values if yes the crate gets that list index assigned instead of making a new one. This allows to use a single list for multiple different crates as long as they share the same values but if its a different value the crate gets a new list assigned that it will share again with all crates with the same values.

MNarath1 commented:
off the matlab course i made a year ago was no waste after all but making this assoziative list took longer then what i initially expected to be honest.

MNarath1 committed:
type in comment oops

MNarath1 committed:
better unique value and some typo fixes

MNarath1 committed:
just swapping around some checks

MNarath1 committed:
ooops

MNarath1 committed:
removes unneeded check

park66665 commented on code/game/objects/structures/crates_lockers/crates.dm:
```diff
@@ -23,9 +23,16 @@
 
 /obj/structure/closet/crate/Initialize()
 	. = ..()
+	if(!door_anim_time == 0&&crateexeption == FALSE)
```
```suggestion
	if(!door_anim_time == 0 && !animation_math["[door_anim_squish]-[door_anim_angle]-[door_hinge]-[door_anim_time]-[radius_2]-[open_angle]"])
```

Proc calls aren't costless. You gotta save time from proc calls that won't do anything but checks.

MNarath1 replied:
I implemented a check for the current code that basically serves that purpose

park66665 commented on code/game/objects/structures/crates_lockers/crates.dm:
```diff
@@ -67,10 +74,9 @@
-		var/matrix/M = get_door_transform(azimuth_angle, polar_angle, radius_cr)
+		var/crateanim_1 = animation_math[anindex][closing ? num_steps + 1 - I : I + 1]
+		var/crateanim_2 = animation_math[anindex][closing ? 2*num_steps + 1 - I : num_steps+ I + 1]
+		var/matrix/M = get_door_transform(crateanim_1, crateanim_2)
```
```suggestion
		var/list/animation_math_list = animation_math["[door_anim_squish]-[door_anim_angle]-[door_hinge]-[door_anim_time]-[radius_2]-[open_angle]"]
		var/matrix/M = get_door_transform(animation_math_list[closing ? num_steps + 1 - I : I + 1], animation_math_list[closing ? 2 * num_steps + 1 - I : num_steps + I + 1])
```

Please let me know if `open_angle` has to change when animation parameters change; if `open_angle` is a replacement for a number that wouldn't change when animation parameters change, you have to make it a define.

MNarath1 replied:
open angle defines the azimuth angle when the crate is open so having the possibility to change it for individual crates is needed for anything crate with another orientation then the current ones like i plan to also animate the coffins later for those i will need to change it to 0.

MNarath1 replied:
I decided to use door_anim_squish instead of open_angle as it is a unused var for most crates(except critter crate)

park66665 replied:
You don't "repurpose" variables.

park66665 commented on code/game/objects/structures/crates_lockers/crates.dm:
```diff
@@ -87,13 +93,38 @@
+The second list is animation_index this list is an associative list where special identify values are saved that are used to determine if there is the need for another list in animation_math or if
+the current lists are sufficent and it is used to get the right index for animation_math for any possible type of crate.
+Next is anindex this variable is just used to save the index for a crate so that you only need to search for the index once.
+The last one is the for loop it simply uses all given indexes to save the needed values to the lists inside animation_math if a new list is required*/
```
```suggestion
```

Obsolete documentation

park66665 commented on code/game/objects/structures/crates_lockers/crates.dm:
```diff
@@ -87,13 +93,38 @@
+			var/azimuth_angle = angle_1 >= 90 ? open_angle : 0
+			var/radius_cr = angle_1 >= 90 ? radius_2 : 1
+			animation_math[anindex][I+1] = -sin(polar_angle)*sin(azimuth_angle)*radius_cr
+			animation_math[anindex][num_steps_1+I+1] = radius_cr*cos(azimuth_angle)*sin(polar_angle)
```
```suggestion
	var/num_steps_1 = door_anim_time / world.tick_lag
	var/list/new_animation_math_sublist[num_steps_1 * 2 + 1]
	animation_math["[door_anim_squish]-[door_anim_angle]-[door_hinge]-[door_anim_time]-[radius_2]-[open_angle]"] = new_animation_math_sublist
	for(var/I in 0 to num_steps_1) //loop to save the animation values into the lists
		var/angle_1 = I == 0 ? 0 : door_anim_angle * (I / num_steps_1)
		var/polar_angle = abs(arcsin(cos(angle_1)))
		var/azimuth_angle = angle_1 >= 90 ? open_angle : 0
		var/radius_cr = angle_1 >= 90 ? radius_2 : 1
		new_animation_math_sublist[I+1] = -sin(polar_angle) * sin(azimuth_angle) * radius_cr
		new_animation_math_sublist[num_steps_1+I+1] = cos(azimuth_angle) * sin(polar_angle) * radius_cr
```

At the suggested rewrite, the proc *will* do the pre-calculation and save it to the `animation_math`.

MNarath1 replied:
sadly does not seem to work only ends in a runtime with bad index i would love to save the index directly into the list index but i do not know if dm actually supports that at all reason why i went with that big workaround anyways

MNarath1 replied:
so yes this should work like in http://www.byond.com/docs/ref/#/list/associations but it does not for some reason no idea if list associations like that are broken at the current byond version

park66665 replied:
Not sure assoc lists are broken in recent versions, but the currently recommended version is 1536

park66665 commented on code/game/objects/structures/crates_lockers/crates.dm:
```diff
@@ -23,9 +23,16 @@
+	var/static/list/animation_math //List with lists inside that contain all values needed for an animation
+	var/static/list/animation_index //associative list to get the index for the animation_math list that gets saved to anindex
+	var/anindex //index from animation_index for animation_math
+	var/crateexeption = FALSE //Set to true if the crate does not use the normal animation code and does not need a list
```
```suggestion
	var/static/list/animation_math //assoc list of lists which contain all values needed for an animation
```

MNarath1 replied:
reason for why i even went with a second list was cause i could not find a way to target  a list inside a list for things like list.len() or list.find()

MNarath1 replied:
Still seems like i need to have a second list to get the index the direct association with the unique value does not seem to work

park66665 commented on code/game/objects/structures/crates_lockers/crates/critter.dm:
```diff
@@ -16,6 +16,7 @@
 	door_hinge = 5.5
 	door_anim_angle = 90
 	door_anim_squish = 0.35
+	crateexeption = TRUE
```
```suggestion
```

MNarath1 replied:
This is just here to make sure critter crate does not get a list as it is using a snowflake code instead

park66665 reviewed (changes requested):
This makes your code mostly sane, and now it seems that the code can handle the pre-calculation of matrices. Please consider implementing it.

Also let me know if my rewrite fails to compile since the suggestion was made on GitHub only.

MNarath1 commented:
> This makes your code mostly sane, and now it seems that the code can handle the pre-calculation of matrices. Please consider implementing it.
> 
> 
> 
> Also let me know if my rewrite fails to compile since the suggestion was made on GitHub only.

- While it does not fail to compile it results in a runtime with bad index even trough it should conform to associate list refs

MNarath1 commented:
and it always fails at the unique value as index so yeah i am not sure its possible to use this methode currently might be broken

MNarath1 commented:
I even tried to use a normal string without any values as assoziative index but that also failed so its not because its actual values that re used like strings it has to do with something else

MNarath1 commented:
I am definitly going to use the sublist define cause its usefull af but i do not think i can get this to work without define the index on another list and then get the list index with find

MNarath1 commented:
@park66665 while defining the index directly as a unique value does not seem to work i can implement some of your other exceptions sadly this still needs the animation_index list to get the index

MNarath1 committed:
implements some suggestions of park and removes some redunant codelines

Implements some suggestions of Park sadly associating the unique value as the index for the list does not seem to work despite it beeing in byond ref like that so i still need two static lists despite just a single one

MNarath1 commented:
I was able to implement some stuff but sadly not everything using a direct association for the index with the unique value does not seem to work for the reason trough i have no idea at all why but at least its better now

MNarath1 commented:
@park66665 If you have any ideas how to possible still implement a direct association with the unique value as index please tell me cause i am out of ideas here looking at the documentation it should work but it clearly does not

park66665 commented:
Don't scream "it runtimes!!". Post the runtimes instead.

MNarath1 commented:
> Don't scream "it runtimes!!". Post the runtimes instead.

The runtime in question was "bad index"

park66665 commented:
The error said nothing but "bad index"? And, did you replace
`animation_math[anindex]` with `animation_math_list` properly?

MNarath1 commented:
........nevermind this is the exact same syntax i used before why the fuck is it working now i tried like 5 times yesterday but then again i also did it like 3 am in the morning

park66665 commented:
I still don't know what is working (Is it my suggested rewrite? Or is it
your changes?) but coding in such late time like 3 am either makes you two
times more productive or sloppy, indeed

MNarath1 commented:
> I still don't know what is working (Is it my suggested rewrite? Or is it your changes?) but coding in such late time like 3 am either makes you two times more productive or sloppy, indeed

I got it to work now with some small changes from your suggested rewrite just a bit on how the index is structured and when the sublist is written into the animation_math list or when its read out in animate_door as the animation_math_list only needs to be var defined once so i put that infront of the for loop thanks for the help

MNarath1 commented:
why exactly it did not work before is still beyond me i have no idea this is how i made it before too but it works now you know i will not question this further

MNarath1 commented:
i think i know now the original rewrite  had no new/list define without that it went  to bad index i included a check in my changes that checked beforehand for this and would call new/list() for animation math if animation_math would be null

MNarath1 committed:
actual directly associated list index now

It actually works a directly associated list this took too long but lesson learned do not try to debug at 3 am

MNarath1 committed:
remove exception var and overrides crate critter animation_list proc

MNarath1 committed:
Includes critter crates into the pre calculated assoc list

park66665 commented:
1. You need more punctuation. Is your every punctuation key (like period(.) and comma(,)) broken?
2. It is great to hear that the code is sailing smoothly now

MNarath1 commented:
> 1. You need more punctuation. Is your every punctuation key (like period(.) and comma(,)) broken?
> 2. It is great to hear that the code is sailing smoothly now

You mean in my description ah yes i suppose i should use punctuations a bit more  and thanks for the help here

MNarath1 committed:
change in comment

MNarath1 commented:
I just closed and open it to run the tests again it has been a while after all

park66665 commented on code/game/objects/structures/crates_lockers/crates.dm:
```diff
@@ -87,13 +91,25 @@
-		M.Multiply(matrix(1, -sin(polar_angle)*sin(azimuth_angle)* radius_cr, 0, 0, radius_cr*cos(azimuth_angle)*sin(polar_angle), 0))
+		M.Multiply(matrix(1, crateanim_1, 0, 0, crateanim_2, 0))
 		M.Translate(0, door_hinge)
 		return M
```
This portion is indented wrongly: there are two tabs

MNarath1 replied:
No idea how i got that in but i am on it

park66665 commented on code/game/objects/structures/crates_lockers/crates.dm:
```diff
@@ -21,11 +21,17 @@
 	open_sound_volume = 35
 	close_sound_volume = 50
 	drag_slowdown = 0
+	door_anim_squish = 138 //in this context the azimuth angle for over 90 degree
```
You don't "re-use" variables. Make a new variable, like `azimuth_angle_2`

MNarath1 replied:
ok i defined a new var instead

park66665 commented on code/game/objects/structures/crates_lockers/crates.dm:
```diff
@@ -21,11 +21,17 @@
 	. = ..()
+	if(animation_math == null) //checks if there is already a list for animation_math if not creates one to avoid runtimes
+		animation_math = new/list()
+	if(!door_anim_time == 0 && !animation_math["[door_anim_time]-[door_anim_angle]-[door_anim_squish]-[radius_2]"])
```
You missed `door_hinge`.

MNarath1 replied:
oops sorry about that

park66665 reviewed (changes requested):
Those little code problems(?) should be fixed. The animation looks okay, so when those are addressed, this PR is good to go.

MNarath1 committed:
implementing requested changes

MNarath1 commented:
@park66665 ok i implemented the changes in a new commit

park66665 reviewed (approved):
Did a final testing; `animation_list()` is only called thrice as intended; animation still looks okay.
